File: src/openrgb_flowers/effects/petal_geometry.py
"""Harmonic petal geometry and polar boundary calculations."""
from __future__ import annotations
import numpy as np
from openrgb_flowers.math.easing import EasingFunctions


class PetalGeometry:
    """Calculates polar petal boundaries and radial intensity falloffs."""

    @staticmethod
    def compute_petal_radii(
        angles: np.ndarray,
        base_radius: float,
        petal_count: int,
        petal_depth: float,
        rotation: float,
    ) -> np.ndarray:
        """Calculates boundary radius for each angle based on harmonic lobes.

        R(theta) = base_radius * (1.0 + petal_depth * cos(lobes * theta + rotation))
        """
        harmonic = EasingFunctions.petal_harmonic(angles, petal_count, petal_depth, rotation)
        return np.maximum(0.001, base_radius * harmonic)

    @staticmethod
    def compute_falloff(
        distances: np.ndarray,
        boundary_radii: np.ndarray,
        edge_softness: float = 0.25,
    ) -> np.ndarray:
        """Calculates smoothstep edge falloff: 1.0 inside flower, transitioning to 0.0 outside."""
        rel_dist = distances / boundary_radii
        # Smooth falloff near the petal boundary
        inner_edge = max(0.0, 1.0 - edge_softness)
        outer_edge = 1.0 + edge_softness * 0.5
        falloff = 1.0 - EasingFunctions.smoothstep(inner_edge, outer_edge, rel_dist)
        return falloff

File: src/effects/petal_geometry.cpp
#include "openrgb_flowers/effects/petal_geometry.hpp"
#include "openrgb_flowers/math/easing.hpp"

#include <algorithm>

namespace openrgb_flowers::effects {

float PetalGeometry::compute_petal_radius(
    float angle,
    float base_radius,
    int petal_count,
    float petal_depth,
    float rotation
) {
    float harmonic = math::EasingFunctions::petal_harmonic(angle, petal_count, petal_depth, rotation);
    return std::max(0.001f, base_radius * harmonic);
}

float PetalGeometry::compute_falloff(
    float distance,
    float boundary_radius,
    float edge_softness
) {
    float boundary = std::max(1e-5f, boundary_radius);
    float rel_dist = distance / boundary;
    float inner_edge = std::max(0.0f, 1.0f - edge_softness);
    float outer_edge = 1.0f + edge_softness * 0.5f;
    return 1.0f - math::EasingFunctions::smoothstep(inner_edge, outer_edge, rel_dist);
}

void PetalGeometry::compute_petal_radii_batch(
    const float* angles,
    float* out_radii,
    size_t count,
    float base_radius,
    int petal_count,
    float petal_depth,
    float rotation
) {
    for (size_t i = 0; i < count; ++i) {
        out_radii[i] = compute_petal_radius(angles[i], base_radius, petal_count, petal_depth, rotation);
    }
}

void PetalGeometry::compute_falloff_batch(
    const float* distances,
    const float* boundary_radii,
    float* out_falloffs,
    size_t count,
    float edge_softness
) {
    for (size_t i = 0; i < count; ++i) {
        out_falloffs[i] = compute_falloff(distances[i], boundary_radii[i], edge_softness);
    }
}

} // namespace openrgb_flowers::effects
